// ldap-ts-client/dist/app.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const dotenv_1 = require("dotenv");
dotenv_1.config();
const index_1 = require("./index");
(async () => {
    var _a, _b, _c;
    const base = "DC=ki,DC=local";
    const options = {
        ldapServerUrl: (_a = process.env.AD_URI) !== null && _a !== void 0 ? _a : "",
        user: (_b = process.env.AD_USER) !== null && _b !== void 0 ? _b : "",
        pass: (_c = process.env.AD_Pass) !== null && _c !== void 0 ? _c : "",
    };
    const client = new index_1.Client(options);
    const data1 = await client.queryAttributes({
        base,
        attributes: ["cn"],
        options: {
            filter: "(&(objectClass=user)(cn=*))",
            scope: "sub",
            paged: true,
        },
    });
    console.log(`File: app.ts,`, `Line: 17 => `, data1.length);
    // const data2 = await client.queryAttributes<User>({
    //   base,
    //   attributes: ["*", "USNIntersite"],
    //   options: {
    //     filter: "(&(objectClass=user)(cn=*))",
    //     scope: "sub",
    //     paged: true,
    //   },
    // });
    // console.log(`File: app.ts,`, `Line: 35 => `, data2.length);
    // const delResult = await client.del({
    //   dn: "CN=testUser2,OU=Users,OU=KII,DC=ki,DC=local",
    // });
    // console.log(`File: app.ts,`, `Line: 18 => `, delResult);
    // const { value } = await client.extendedOp({
    //   oid: "1.3.6.1.4.1.4203.1.11.3",
    //   value: "",
    // });
    // console.log(`File: app.ts,`, `Line: 18 => `, value);
    // const modifyDnResult = await client.modifyDn({
    //   dn: "CN=testUser3,OU=Users,OU=KII,DC=ki,DC=local",
    //   newDn: "CN=testUser4,OU=Users,OU=KII,DC=ki,DC=local",
    // });
    // console.log(`File: app.ts,`, `Line: 18 => `, modifyDnResult);
    // const entry = {
    //   cn: "testUser2",
    //   sn: "testUser2",
    //   objectClass: "organizationalPerson",
    // };
    // const addResult = await client.add<User>({
    //   entry,
    //   dn: "CN=testUser2,OU=Users,OU=KII,DC=ki,DC=local",
    // });
    // console.log(addResult);
    // const modifyAttributeResult = await client.modifyAttribute<User>({
    //   dn: "CN=testUser2,OU=Users,OU=KII,DC=ki,DC=local",
    //   changes: [
    //     {
    //       operation: "add",
    //       modification: {
    //         mail: "mymail@mydomain.com",
    //       },
    //     },
    //   ],
    // });
    // console.log(`File: app.ts,`, `Line: 53 => `, modifyAttributeResult);
    // const compared = await client.compare<User>({
    //   dn: "CN=testUser2,OU=Users,OU=KII,DC=ki,DC=local",
    //   field: {
    //     cn: "testUser2",
    //   },
    // });
    // console.log(`File: app.ts,`, `Line: 22 => `, compared);
    // const data = await client.queryAttributes<User>({
    //   base,
    //   attributes: ["cn"],
    //   options: {
    //     filter: "(&(objectClass=user)(cn=*))",
    //     scope: "sub",
    //     paged: true,
    //   },
    // });
    // console.log(`File: app.ts,`, `Line: 95 => `, data.length);
    client.unbind();
})();
